### PR TITLE
Expose connection class object to NetworkTemplate

### PR DESCRIPTION
**Pull request by GomathiselviS**
Signed-off-by: GomathiselviS <gomathiselvi@gmail.com>
Depends-On: https://github.com/ansible-collections/ansible.netcommon/pull/221
##### SUMMARY

Inorder to get the network os version, connection class object needs to be accessed inside the rm_template file.
To achieve this, a base class has been added - RmbaseEngine which handles the connection to the device. This class is  inherited by ResourceModule and NetworkTemplate class.

##### ISSUE TYPE

- Bugfix Pull Request

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by GomathiselviS** [comment]
recheck

**Review by Qalthos — APPROVED** [review]

**Review by ganeshrn — APPROVED** [review]
Thanks for working on it. Changes look fine to me.

**Review by NilashishC — APPROVED** [review]

**Review by NilashishC — COMMENTED** [review]
To organize the classes in a better (more Python-ish) way, we can possibly have the following dir structure:
```
module_utils
-> network
   -> common
      -> rm_base
           -> __init__.py (RMBase)
           -> network_template.py (NetworkTemplate)
           -> resource_module.py (ResourceModule)
```

For backwards compatibility, we can keep the existing `network_template.py` and `resource_module.py` files, remove everything and just add the following respectively in it.
```
from ansible_collections.ansible.netcommon.plugins.module_utils.network.common.rm_base.network_template import (
    NetworkTemplate,
)
from ansible_collections.ansible.netcommon.plugins.module_utils.network.common.rm_base.resource_module import (
    ResourceModule,
)
```

File: plugins/module_utils/network/common/resource_module_base.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+from ansible_collections.ansible.netcommon.plugins.module_utils.network.common.network import (
+    get_resource_connection,
+)
+
+
+class RmEngineBase(object):  # pylint: disable=R0902
+    """ Base class for Network Resource Modules
+    """
+
+    ACTION_STATES = ["merged", "replaced", "overridden", "deleted"]
```
Comment (NilashishC) [inline plugins/module_utils/network/common/resource_module_base.py:14]:
```suggestion
    ACTION_STATES = ["merged", "replaced", "overridden", "deleted",  "purged"]
```

File: plugins/module_utils/network/common/rm_base/resource_module.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
+from __future__ import absolute_import, division, print_function
```
Comment (NilashishC) [inline plugins/module_utils/network/common/rm_base/resource_module.py:1]: Missing copyright and licensing information.

**Review by NilashishC — CHANGES_REQUESTED** [review]

File: tox.ini [hunk 3]
```diff
@@ -28,4 +28,4 @@ show-source = True
 ignore = E123,E125,E402,W503
 max-line-length = 160
 builtins = _
-exclude = .git,.mypy_cache,.tox,tests/unit/compat/,tests/output/
+exclude = .git,.mypy_cache,.tox,tests/unit/compat/,tests/output/,plugins/module_utils/network/common/resource_module.py,plugins/module_utils/network/common/network_template.py
```
Comment (NilashishC) [inline tox.ini:31]: Why do we need this exclude?

File: plugins/module_utils/network/common/network_template.py [hunk 4]
```diff
@@ -2,110 +2,6 @@
 
 __metaclass__ = type
 
-import re
-from copy import deepcopy
-
-from ansible_collections.ansible.netcommon.plugins.module_utils.network.common.utils import (
-    Template,
-    dict_merge,
+from ansible_collections.ansible.netcommon.plugins.module_utils.network.common.rm_base.network_template import (
```
Comment (NilashishC) [inline plugins/module_utils/network/common/network_template.py:5]:
```
from ansible_collections.ansible.netcommon.plugins.module_utils.network.common.rm_base.network_template import (  # noqa: F401
    NetworkTemplate,
)
```

File: plugins/module_utils/network/common/resource_module.py [hunk 5]
```diff
@@ -2,164 +2,6 @@
 
 __metaclass__ = type
 
-from copy import deepcopy
-from ansible_collections.ansible.netcommon.plugins.module_utils.network.common.utils import (
-    remove_empties,
-    to_list,
-    get_from_dict,
+from ansible_collections.ansible.netcommon.plugins.module_utils.network.common.rm_base.resource_module import (
```
Comment (NilashishC) [inline plugins/module_utils/network/common/resource_module.py:5]:
```
from ansible_collections.ansible.netcommon.plugins.module_utils.network.common.rm_base.resource_module import (  # noqa: F401
    ResourceModule,
)
```

File: tox.ini (hunk 3 above)
Comment (NilashishC) [reply tox.ini:31]: Okay, we can just skip the F401 checks on those two imports. I have commented on those files.

**Comment by Qalthos** [comment]
recheck

**Review by NilashishC — APPROVED** [review]

**Comment by GomathiselviS** [comment]
recheck

**Comment by rohitthakur2590** [comment]
recheck

**Pull request merged**